dartagnan/src/main/java/com/dat3m/dartagnan/wmm/Wmm.java

```diff
@@ -260,71 +209,115 @@ private void simplifyAssociatives(Class<? extends Definition> cls, BiFunction<Re
         }
     }
 
-    private Definition basicDefinition(String name) {
-        Relation r = newRelation(name);
-        return switch (name) {
+    public void removeUnconstrainedRelations() {
+        // TODO: Currently removes all axiom-irrelevant relations and thus may remove
+        //  other relations part of other constraints.
+        final DependencyCollector collector = new DependencyCollector();
+        getAxioms().forEach(ax -> ax.accept(collector));
+        final Set<Relation> relevantRelations = new HashSet<>(collector.collectedRelations);
+        Wmm.ANARCHIC_CORE_RELATIONS.forEach(n -> relevantRelations.add(getRelation(n)));
+
+        for (Constraint c : List.copyOf(getConstraints())) {
+            if (!relevantRelations.containsAll(c.getConstrainedRelations())) {
+                removeConstraint(c);
+            }
+        }
+
+        for (Relation rel : Set.copyOf(getRelations())) {
+            if (!relevantRelations.contains(rel)) {
+                deleteRelation(rel);
+            }
+        }
+}
+
+    private final static class DependencyCollector implements Constraint.Visitor<Void> {
+        private final Set<Relation> collectedRelations = new HashSet<>();
+        @Override
+        public Void visitConstraint(Constraint constraint) {
+            for (Relation rel :  constraint.getConstrainedRelations()) {
+                if (collectedRelations.add(rel)) {
+                    rel.getDefinition().accept(this);
+                }
+            }
+            return null;
+        }
+    }
+
+    private Relation makePredefinedRelation(String name) {
+        /*
+            WARNING: The code has possibly unexpected behaviour:
+             - If a relation like `ext` is already defined (different from the standard definition),
+               constructing e.g. `rfe` will refer to this new definition.
+             - When constructing derived relations like fr, the intermediately defined relations like rf^-1
+               will get constructed even if they already exist in the model.
+             TODO: Clarify what the intended behaviour should be.
+         */
```

Comment:
Is it because we internally hardcode the expected definition of things like `rfe`?
I think we should not do this and make use of the explicit definition in the cat file

Reply:
Yes, it's because of hardcoded definitions.
I fully agree that we should define all derived relations in the cat-file, however, that means we should have proper support for includes.

Also, the issue would still exist if you programatically tried to generate a `Wmm` rather than using the parser, but this is a rather rare occurence (though the current code for creating biases does it).